activerecord/lib/active_record/relation.rb

```diff
@@ -699,6 +701,13 @@ def skip_query_cache_if_necessary
         end
       end
 
+      def retain_positionals
+        unless @offsets.empty?
+          unpositioned = @records.reject { |r| @offsets.values.include?(r) }.each
```

Comment:
I really like this use of an explicit Enumerator!

But I think it's probably worth assuming that `@records` will be fairly large here, especially compared to `@offsets`, so if possible we should avoid [explicitly] iterating it.

How about something like:

```ruby
if @offsets.all? { |offset, record| @records[offset] == record }
  @offsets.each { |offset, record| @records[offset] = record }
else
  @records -= @offsets.values
  @offsets.sort.each { |offset, record| @records[offset, 0] = record }
end
```

* the `all?` will subsume the `empty?` check, and also skip doing the expensive work if everything is already in the right place
* (but we'll still swap in the previously-loaded instances, so that behaviour's predictable regardless of the records' "natural" order)
* `Array#-` should be the fastest [simple] way of doing the exclusion operation we want (because the looping is all in C)
* I haven't checked, but I *assume* insertions will shift large blocks of array content around in a handful of low-level memory operations.. it's uglier, but it seems reasonable to expect it'll be faster

Reply:
I hadn't thought of the difference in size—this makes a lot of sense! It's unclear to me what the purpose of the line in the if block serves; if the `all?` statement has been satisfied, shouldn't all records already match the offsets? Why do they need to be swapped again? Is this a case of shallow equals followed by setting a pointer?

Reply:
Yes, exactly. They're objects representing the right database rows, but they're freshly constructed object instances -- semantically equal, but not identity-equal. The previously-loaded ones might have already fetched and cached a child association, for example, so I think we're better off keeping them.